### PR TITLE
Add detection for Sequoia problems with shortcuts

### PR DESCRIPTION
**Pull request by alanflex**
This PR adds extra checks in `SRShortcutControl` to detect a shortcut which won't work on Sequoia.

The first time a shortcut is attempted which won't work on Sequoia, an `NSAlert` is displayed (not yet copyedited or localised in this PR).  The alert won't be displayed again for the lifetime of the control, but the shortcut will continued to be denied as long as it doesn't qualify.

Have tried to match the code and documentation style of the existing component, but most of it's checks for validity are handled by setting flags or delegate callbacks so this is, by definition, a little different.

### CODE REVIEW

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 1]
```diff
@@ -891,6 +897,13 @@ - (BOOL)areModifierFlagsValid:(NSEventModifierFlags)aModifierFlags forKeyCode:(S
     return allowModifierFlags;
 }
 
+- (BOOL)isRunningOnSequoia {
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m]: Brace on newline

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 2]
```diff
@@ -891,6 +897,13 @@ - (BOOL)areModifierFlagsValid:(NSEventModifierFlags)aModifierFlags forKeyCode:(S
...
+    if( @available(macOS 15.0, *) )
+        return YES;
+    else
+        return NO;
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m]: Add braces around if and else, change if whitespace to match elsewhere.

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 3]
```diff
@@ -913,7 +926,11 @@ - (BOOL)areModifierFlagsAllowed:(NSEventModifierFlags)aModifierFlags forKeyCode:
 #pragma clang diagnostic pop
 
     os_activity_initiate("-[SRRecorderControl areModifierFlagsAllowed:forKeyCode:]", OS_ACTIVITY_FLAG_IF_NONE_PRESENT, ^{
-        if ((aModifierFlags == 0 && !self.allowsEmptyModifierFlags) ||
+        if( [self isRunningOnSequoia] && ![SRShortcut sequoiaValidModifiers:aModifierFlags])
+        {
+            allowModifierFlags = NO;
+        }
+        else if ((aModifierFlags == 0 && !self.allowsEmptyModifierFlags) ||
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m]: Braces on same line as if and else if. Whitespace of if.

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 4]
```diff
@@ -1045,7 +1072,12 @@ - (BOOL)canEndRecordingWithObjectValue:(nullable SRShortcut *)aShortcut
 #pragma clang diagnostic pop
 
     os_activity_initiate("-[SRRecorderControl canEndRecordingWithObjectValue:]", OS_ACTIVITY_FLAG_DEFAULT, ^{
-        if ([self areModifierFlagsValid:aShortcut.modifierFlags forKeyCode:aShortcut.keyCode])
+        if( [self isRunningOnSequoia] && ![SRShortcut sequoiaValidModifiers:aShortcut.modifierFlags] )
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m]: Whitespace

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 5]
```diff
@@ -1045,7 +1072,12 @@ - (BOOL)canEndRecordingWithObjectValue:(nullable SRShortcut *)aShortcut
...
+        {
+            os_log_debug(OS_LOG_DEFAULT, "Modifiers %lu rejected on Sequoia", aShortcut.modifierFlags);
+            result = NO;
+        }
+        else if ([self areModifierFlagsValid:aShortcut.modifierFlags forKeyCode:aShortcut.keyCode])
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m:1086]: Braces on same line as if and else if.

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 6]
```diff
@@ -1766,6 +1798,9 @@ - (BOOL)performKeyEquivalent:(NSEvent *)anEvent
                 {
                     // Do not end editing and allow the client to make another attempt.
                     [self playAlert];
+
+                    if( [self isRunningOnSequoia] && ![newObjectValue isValidOnSequoia] && !_hasSequoiaAlertBeenShown)
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m]: Whitespace

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 7]
```diff
@@ -928,6 +945,16 @@ - (void)playAlert
     NSBeep();
 }
 
+- (void)showSequoiaAlert
+{
+    NSAlert *sequoiaAlert = [[NSAlert alloc] init];
+    [sequoiaAlert setMessageText:SRLoc(@"Shortcut not valid on macOS Sequoia")];
+    [sequoiaAlert setInformativeText:SRLoc(@"Sequoia cannot use a shortcut with just opt or shift as modifiers.  Opt and shift are valid, but must be combined with cmd or ctrl.")];
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m]: Wording suggestions:

- Shortcut not valid
- Shortcuts must include the command or control key. Shortcuts with only option or shift are not allowed by macOS.

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 8]
```diff
@@ -928,6 +945,16 @@ - (void)playAlert
     NSBeep();
 }
 
+- (void)showSequoiaAlert
+{
+    NSAlert *sequoiaAlert = [[NSAlert alloc] init];
```
Comment (ksuther) [inline Sources/ShortcutRecorder/SRRecorderControl.m:955]: Add newline below

**Review by ksuther — COMMENTED** [review]
Left some code style nitpicks and wording suggestions.

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 8 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m:955]: 👍🏻

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 7 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: 👍🏻

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 6 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: 👍🏻

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 2 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: Happy to change but the reason I took all the braces out is they don't seem to be used elsewhere for single lines.  Look for example at `-[SRShortcut isEqualToDictionary]` or `-[SRRecorderControl flagsChanged]`

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 3 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: Again happy to change it, but if you look at, say `-[SRRecorderControl performKeyEquivalent:]` they seem to prefer all braces on a new line ?

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 4 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: 👍🏻

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 5 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m:1086]: (see elsewhere, that doesn't seem to be how its done elsewhere in the code)

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 1 above)
Comment (alanflex) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: dang I missed one.

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 3 above)
Comment (ksuther) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: Oh oops, never mind then :)

File: Sources/ShortcutRecorder/SRRecorderControl.m (hunk 2 above)
Comment (ksuther) [reply Sources/ShortcutRecorder/SRRecorderControl.m]: I would argue this is a style worth ignoring, as single line if statements without braces are just asking for goto fail style bugs. I see other places in the code that aren't doing this too.

**Review by ksuther — APPROVED** [review]
I still think one line if/elses are bad, but otherwise looks good to me :)

**Comment by alanflex** [comment]
@ksuther I 100% agree but I was trying to match the crazy style :-)

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 9]
```diff
@@ -891,6 +897,18 @@ - (BOOL)areModifierFlagsValid:(NSEventModifierFlags)aModifierFlags forKeyCode:(S
     return allowModifierFlags;
 }
 
+- (BOOL)isRunningOnSequoia
```
Comment (vendruscolo) [inline Sources/ShortcutRecorder/SRRecorderControl.m:900]: Nit: I would name this differently as it's likely Apple won't change the behavior in macOS 16. Some names to consider:
* isRunningOnSequoiaOrMore
* isAtLeastMacOS15
* needsCtrlOrCmdModifiers

(last one wins imo)

File: Sources/ShortcutRecorder/SRRecorderControl.m [hunk 10]
```diff
@@ -928,6 +950,17 @@ - (void)playAlert
     NSBeep();
 }
 
+- (void)showSequoiaAlert
+{
+    NSAlert *sequoiaAlert = [[NSAlert alloc] init];
+    
+    [sequoiaAlert setMessageText:SRLoc(@"Shortcut not valid")];
+    [sequoiaAlert setInformativeText:SRLoc(@"Shortcuts must include the command or control key. Shortcuts with only option or shift are not allowed by macOS.")];
```
Comment (vendruscolo) [inline Sources/ShortcutRecorder/SRRecorderControl.m:958]: The HIG capitalize key names (https://developer.apple.com/design/human-interface-guidelines/keyboards)
We should match so it's Command, Control, Option, and Shift.

File: Sources/ShortcutRecorder/SRShortcut.m [hunk 11]
```diff
@@ -222,6 +231,13 @@ - (instancetype)initWithCode:(SRKeyCode)aKeyCode
     return d;
 }
 
+// A shortcut is invalid on Sequoia if it contains ONLY shift or option modifiers.
+// no modifiers is fine, but if any modifer is present, it must contain cmd and/or ctrl
+// https://developer.apple.com/forums/thread/763878?answerId=804374022#804374022
+- (BOOL) isValidOnSequoia
```
Comment (vendruscolo) [inline Sources/ShortcutRecorder/SRShortcut.m:237]: Nit: space after `(BOOL)`

File: Sources/ShortcutRecorder/SRShortcut.m [hunk 12]
```diff
@@ -203,6 +203,15 @@ - (instancetype)initWithCode:(SRKeyCode)aKeyCode
     return self;
 }
 
+#pragma mark Non-initializer class methods
+
++ (BOOL)sequoiaValidModifiers:(NSEventModifierFlags)modifierFlags
```
Comment (vendruscolo) [inline Sources/ShortcutRecorder/SRShortcut.m:208]: I wouldn't name it with 'Sequoia' in it, as it'll be out of date next year.
I wonder if we bring the `if @available()` check in here for simplicity? So we have a single entry point that is `validModifiers:`?

**Pull request merged**